File: C#/Memento/Memento.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Memento.Memento
{
    class Memento : IMemento
    {
        private int state;
        private string name;
        private Originator originator;

        public Memento(Originator originator, int state, string name)
        {
            Console.WriteLine("MEMENTO: Creating memento");
            this.originator = originator;
            this.state = state;            
            this.name = name;
        }

        //restoring orignator state from memento
        public void RestoreMemento()
        {
            Console.WriteLine("MEMENTO: Restoring state");
            originator.SetState(state, name);
        }               

        public override string ToString()
        {
            Console.Write("MEMENTO: Getting memento state: ");
            return state.ToString();
        }
    }
}

File: python/Memento/Memento.py
from IMemento import IMemento
import Originator
import string
import datetime

class Memento(IMemento):

    def __init__(self, originator: Originator, state: int, name: string) -> None:
        print("MEMENTO: Creating memento")
        self.originator = originator
        self.state = state
        self.name = name
        self.date = datetime.date.today()

    def RestoreMemento(self) -> None:
        print(f"MEMENTO: Restoring state from {self.date}")
        self.originator.SetState(self.state, self.name)

    def ShowMemento(self) -> string:
        return f"State: {str(self.state)}"